bin/propolis-utils/src/bin/cpuid-gen.rs

```diff
@@ -261,15 +262,116 @@ fn print_toml(results: &BTreeMap<CpuidKey, Cpuid>) {
     }
 }
 
+fn print_json(results: &BTreeMap<CpuidKey, Cpuid>) {
+    use propolis_api_types::instance_spec::components::board::{
+        Cpuid, CpuidEntry,
+    };
+
+    let vendor = {
+        use propolis_api_types::instance_spec::{CpuidValues, CpuidVendor};
+        match results.get(&CpuidKey::Leaf(0)) {
+            None => {
+                eprintln!("no result for leaf 0, selecting default vendor");
+                CpuidVendor::Amd
+            }
+            Some(values) => {
+                let values = CpuidValues {
+                    eax: values.eax,
+                    ebx: values.ebx,
+                    ecx: values.ecx,
+                    edx: values.edx,
+                };
+
+                match CpuidVendor::try_from(values) {
+                    Err(_) => {
+                        eprintln!(
+                            "vendor in leaf 0 values ({values:?}) \
+                        not recognized, selecting default vendor"
+                        );
+                        CpuidVendor::Amd
+                    }
+                    Ok(v) => v,
+                }
+            }
+        }
+    };
+
+    let cpuid = Cpuid {
+        entries: results
+            .iter()
+            .filter_map(|(key, value)| {
+                // propolis-server will complain if it receives a CPUID entry
+                // with no subleaf and a CPUID entry with a subleaf for the same
+                // leaf. Avoid emitting these by filtering out subleaf-less
+                // entries for any leaf that has an entry with subleaf data.
+                let (leaf, subleaf) = match key {
+                    CpuidKey::Leaf(leaf) => {
+                        if results.keys().any(|k| {
```

Comment:
If N was more than dozens, at most, I'd say to use `range()` here to constrain the haystack, but it's not worth the effort.

Reply:
We can go one better here: since `CpuidKey`'s `Ord` implementation guarantees that `CpuidKey::Leaf` precedes `CpuidKey::SubLeaf` provided that the leaf indices are equal, we only need to compare to the next entry in the map to see if there's a conflict. I implemented this in a274706.